Flaky testRuntimeOptimizedStagesCorrectness

```
[ERROR] Tests run: 2181, Failures: 1, Errors: 0, Skipped: 90, Time elapsed: 3,146.362 s <<< FAILURE! - in TestSuite
[ERROR] testRuntimeOptimizedStagesCorrectness(com.facebook.presto.hive.TestEventListenerWithExchangeMaterialization)  Time elapsed: 1.398 s  <<< FAILURE!
java.lang.AssertionError: expected [1] but found [0]
	at org.testng.Assert.fail(Assert.java:94)
	at org.testng.Assert.failNotEquals(Assert.java:513)
	at org.testng.Assert.assertEqualsImpl(Assert.java:135)
	at org.testng.Assert.assertEquals(Assert.java:116)
	at org.testng.Assert.assertEquals(Assert.java:389)
	at org.testng.Assert.assertEquals(Assert.java:399)
	at com.facebook.presto.hive.TestEventListenerWithExchangeMaterialization.testRuntimeOptimizedStagesCorrectness(TestEventListenerWithExchangeMaterialization.java:110)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

--- a/presto-hive/src/test/java/com/facebook/presto/hive/TestEventListenerWithExchangeMaterialization.java
+++ b/presto-hive/src/test/java/com/facebook/presto/hive/TestEventListenerWithExchangeMaterialization.java
@@ -23,7 +23,6 @@ import com.facebook.presto.spi.eventlistener.QueryCompletedEvent;
 import com.facebook.presto.spi.eventlistener.QueryCreatedEvent;
 import com.facebook.presto.spi.eventlistener.QueryMetadata;
 import com.facebook.presto.spi.eventlistener.SplitCompletedEvent;
-import com.facebook.presto.testing.MaterializedResult;
 import com.facebook.presto.tests.DistributedQueryRunner;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
@@ -75,6 +74,9 @@ public class TestEventListenerWithExchangeMaterialization
                 Optional.empty());
         queryRunner.installPlugin(new TestingEventListenerPlugin(generatedEvents));
         session = queryRunner.getDefaultSession();
+        generatedEvents.initialize(16);
+        // Wait until all tpch table copying queries finish populating completedEvents.
+        generatedEvents.waitForEvents(180);
     }
 
     @AfterClass(alwaysRun = true)
@@ -86,13 +88,13 @@ public class TestEventListenerWithExchangeMaterialization
         generatedEvents = null;
     }
 
-    private MaterializedResult runQueryAndWaitForEvents(@Language("SQL") String sql, int numEventsExpected)
+    private QueryId runQueryAndWaitForEvents(@Language("SQL") String sql, int numEventsExpected)
             throws Exception
     {
         generatedEvents.initialize(numEventsExpected);
-        MaterializedResult result = queryRunner.execute(session, sql);
-        generatedEvents.waitForEvents(100);
-        return result;
+        QueryId resultId = queryRunner.executeWithQueryId(session, sql).getQueryId();
+        generatedEvents.waitForEvents(600);
+        return resultId;
     }
 
     @Test
@@ -101,12 +103,13 @@ public class TestEventListenerWithExchangeMaterialization
     {
         // We expect one runtime optimized stage: 1.
         int expectedEvents = 2;
-        runQueryAndWaitForEvents("SELECT phone, regionkey FROM nation INNER JOIN supplier ON supplier.nationkey=nation.nationkey", expectedEvents);
+        QueryId queryId = runQueryAndWaitForEvents("SELECT phone, regionkey FROM nation INNER JOIN supplier ON supplier.nationkey=nation.nationkey", expectedEvents);
         QueryCreatedEvent queryCreatedEvent = getOnlyElement(generatedEvents.getQueryCreatedEvents());
         QueryCompletedEvent queryCompletedEvent = getOnlyElement(generatedEvents.getQueryCompletedEvents());
         QueryMetadata queryMetadata = queryCompletedEvent.getMetadata();
         Optional<List<StageId>> runtimeOptimizedStages = queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(new QueryId(queryCreatedEvent.getMetadata().getQueryId())).getRuntimeOptimizedStages();
 
+        assertEquals(queryMetadata.getQueryId(), queryId.toString());
         assertEquals(queryMetadata.getRuntimeOptimizedStages().size(), 1);
         assertEquals(queryMetadata.getRuntimeOptimizedStages().get(0), "1");
         assertTrue(runtimeOptimizedStages.isPresent());

